includes/class-wc-order-product-lookup-helper.php

```diff
@@ -0,0 +1,208 @@
+
+	/**
+	 * Handler for saving an order, keep the lookup table in sync.
+	 *
+	 * @param WC_Order $order Order object.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_new_order( $order ) {
+		global $wpdb;
+
+		$order_items = $order->get_items();
+
+		foreach ( $order_items as $item_id => $item ) {
+			$item_product_id = $item->get_product_id();
+
+			// Skip if this line item doesn't have a product id associated.
+			if ( empty( $item_product_id ) ) {
+				continue;
+			}
+
+			$data = array(
+				'user_email'           => $order->get_billing_email(),
+				'order_id'             => $order->get_id(),
+				'user_id'              => $order->get_customer_id(),
+				'product_id'           => $item_product_id,
+				'parent_order_item_id' => $item_id,
+			);
+
+			$wpdb->insert(
+				$wpdb->prefix . 'woocommerce_order_product_lookup',
+				$data,
+				array(
+					'%s',
+					'%d',
+					'%d',
+					'%d',
+					'%d',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Handler for updating an order, keep the lookup table in sync.
+	 *
+	 * @param int $order_id Order ID.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function handle_update_order( $order_id ) {
+		// Make sure we delete the old order from the lookup database, as it may have its line items updated.
+		self::handle_delete_order( $order_id );
+
+		$order = wc_get_order( $order_id );
+
+		if ( is_a( $order, 'WC_Order' ) ) {
+			self::handle_new_order( $order );
+		}
+	}
+
+	/**
+	 * Check if a product is purchased, given either email or user id.
+	 *
+	 * @param  int        $product_id       Product ID.
+	 * @param  string|int $email_or_user_id Either email or user id.
+	 * @return bool       Whether product is purchased or not.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function is_product_purchased( $product_id, $email_or_user_id ) {
+		global $wpdb;
+
+		if ( is_int( $email_or_user_id ) ) {
+			$orders = $wpdb->get_results( $wpdb->prepare( "
+				SELECT order_id
+				FROM {$wpdb->prefix}woocommerce_order_product_lookup
+				WHERE ( user_id = %d ) AND product_id = %d
+				LIMIT 1
+			", absint( $email_or_user_id ), absint( $product_id ) ) ); // WPCS: cache ok.
+		} else {
+			$orders = $wpdb->get_results( $wpdb->prepare( "
+				SELECT order_id
+				FROM {$wpdb->prefix}woocommerce_order_product_lookup
+				WHERE ( user_email = %s ) AND product_id = %d
+				LIMIT 1
+			", esc_sql( $email_or_user_id ), absint( $product_id ) ) ); // WPCS: cache ok.
+		}
+
+		foreach ( $orders as $order ) {
+			$order = wc_get_order( $order->order_id );
+
+			if ( is_callable( array( $order, 'get_date_paid' ) ) ) {
+				$date_paid = $order->get_date_paid();
+
+				// Every order is marked as paid at most once.
+				if ( ! empty( $date_paid ) ) {
+					return true;
+				}
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Get last order for a given user id.
+	 *
+	 * @param  int      $user_id User ID.
+	 * @return WC_Order Order object.
+	 *
+	 * @since 3.3.0
+	 */
+	public static function get_last_order( $user_id ) {
```

Comment:
Both approaches are viable, but just to point out https://github.com/woocommerce/woocommerce/issues/17087 will make this more performant without a lookup table so we may be able to drop this method.